File: src/c/ch01/1.4/l0/float_precision.c
#include "float_precision.h"
#include <math.h>

bool almost_equal(double x, double y, double eps) {
    return fabs(x - y) < eps;
}

double repeat_add(double value, int times) {
    double s = 0.0;
    for (int i = 0; i < times; ++i) s += value;
    return s;
}

double mix_large_small(double large, double small) {
    return large + small;
}

double sum_naive(const double *xs, size_t n) {
    double s = 0.0;
    for (size_t i = 0; i < n; ++i) s += xs[i];
    return s;
}

File: src/erlang/ch01/1.4/l0/src/float_precision.erl
-module(float_precision).
-export([almost_equal/3,
         repeat_add/2,
         mix_large_small/2,
         sum_naive/1]).

%% almost_equal(X,Y,Eps) -> true if |X-Y| < Eps.
almost_equal(X, Y, Eps) ->
    abs(X - Y) < Eps.

%% repeat_add(Value, Times) -> add Value to 0.0 'Times' times.
repeat_add(Value, Times) when Times =< 0 -> 0.0;
repeat_add(Value, Times) ->
    repeat_add_(Value, Times, 0.0).

repeat_add_(Value, 0, Acc) -> Acc;
repeat_add_(Value, N, Acc) -> repeat_add_(Value, N-1, Acc + Value).

%% Adding tiny to huge may not change the number (limited precision).
mix_large_small(Large, Small) ->
    Large + Small.

%% sum_naive(List) -> naive left-to-right sum (shows accumulated error).
sum_naive(List) ->
    lists:foldl(fun(X, A) -> A + X end, 0.0, List).
